### PR TITLE
✨ Source Amazon Seller Partner: add new stream OrderItems

### PR DESCRIPTION
**Pull request by efimmatytsin**
## What
Add new stream https://developer-docs.amazon.com/sp-api/docs/orders-api-v0-reference#get-ordersv0ordersorderidorderitems

## How
Create stream slice from Orders stream.

## Recommended reading order
1. `x.java`
2. `y.python`

## 🚨 User Impact 🚨
None.

### CODE REVIEW

File: airbyte-integrations/connectors/source-amazon-seller-partner/source_amazon_seller_partner/schemas/OrderItems.json [hunk 1]
```diff
@@ -0,0 +1,269 @@
+{
+    "title": "Order Items",
+    "description": "All order items that were updated after a specified date",
+    "$schema": "http://json-schema.org/draft-07/schema#",
+    "type": "object",
+    "additionalProperties": true,
+    "properties": {
+        "AmazonOrderId": {
+            "type": "string"
```
Comment (marcosmarxm) [inline airbyte-integrations/connectors/source-amazon-seller-partner/source_amazon_seller_partner/schemas/OrderItems.json:9]: Can you add the pattern `["null", "data type"]`? It isn't guarantee to always receive data for each field.

File: airbyte-integrations/connectors/source-amazon-seller-partner/source_amazon_seller_partner/schemas/OrderItems.json [hunk 2]
```diff
@@ -0,0 +1,269 @@
...
+        },
+        "Title": {
+            "type": "string"
+        },
+        "QuantityOrdered": {
+            "type": "integer"
+        },
+        "ProductInfo": {
+            "type": "object",
+            "properties": {
```
Comment (marcosmarxm) [inline airbyte-integrations/connectors/source-amazon-seller-partner/source_amazon_seller_partner/schemas/OrderItems.json:28]: Add additionalProperties: true

File: airbyte-integrations/connectors/source-amazon-seller-partner/source_amazon_seller_partner/source.py [hunk 3]
```diff
@@ -136,7 +137,7 @@ def check_connection(self, logger: AirbyteLogger, config: Mapping[str, Any]) ->
         """
         try:
             stream_kwargs = self._get_stream_kwargs(config)
-            stream_to_check = VendorSalesReports(**stream_kwargs)
+            stream_to_check = Orders(**stream_kwargs)
```
Comment (marcosmarxm) [inline airbyte-integrations/connectors/source-amazon-seller-partner/source_amazon_seller_partner/source.py:140]: Why are you changing this?

File: airbyte-integrations/connectors/source-shopify/source_shopify/utils.py [hunk 4]
```diff
@@ -266,7 +266,7 @@ class EagerlyCachedStreamState:
     It's empty, once the sync has started and is being updated while sync operation takes place,
     It holds the `temporary stream state values` before they are updated to have the opportunity to reuse this state.
     """
-
+cached_state: Dict = {}
```
Comment (marcosmarxm) [inline airbyte-integrations/connectors/source-shopify/source_shopify/utils.py]: Please fix this

File: airbyte-integrations/connectors/source-tiktok-marketing/Dockerfile [hunk 5]
```diff
@@ -32,5 +32,5 @@ COPY source_tiktok_marketing ./source_tiktok_marketing
 ENV AIRBYTE_ENTRYPOINT "python /airbyte/integration_code/main.py"
 ENTRYPOINT ["python", "/airbyte/integration_code/main.py"]
 
-LABEL io.airbyte.version=3.4.0
+LABEL io.airbyte.version=dev1
```
Comment (marcosmarxm) [inline airbyte-integrations/connectors/source-tiktok-marketing/Dockerfile]: Bump following SEMVER. It's a new stream, update to `3.5.0`

**Review by marcosmarxm — CHANGES_REQUESTED** [review]
Missing documentation update and other comments.

File: airbyte-integrations/connectors/source-amazon-seller-partner/source_amazon_seller_partner/source.py (hunk 3 above)
Comment (efimmatytsin) [reply airbyte-integrations/connectors/source-amazon-seller-partner/source_amazon_seller_partner/source.py:140]: hello @marcosmarxm, Im sorry, I forget to explain this. It's not working, only returning 400 client error. We have all permissions in our app, Orders works like a charm.

File: airbyte-integrations/connectors/source-amazon-seller-partner/source_amazon_seller_partner/schemas/OrderItems.json (hunk 2 above)
Comment (efimmatytsin) [reply airbyte-integrations/connectors/source-amazon-seller-partner/source_amazon_seller_partner/schemas/OrderItems.json:28]: ok

File: airbyte-integrations/connectors/source-amazon-seller-partner/source_amazon_seller_partner/schemas/OrderItems.json (hunk 1 above)
Comment (efimmatytsin) [reply airbyte-integrations/connectors/source-amazon-seller-partner/source_amazon_seller_partner/schemas/OrderItems.json:9]: ok

File: airbyte-integrations/connectors/source-amazon-seller-partner/source_amazon_seller_partner/streams.py [hunk 6]
```diff
@@ -103,13 +103,16 @@ def request_params(
...
-            params.update({self.replication_start_date_field: start_date})
+        if self.replication_start_date_field is not None:
+            params.update({self.replication_start_date_field: self._replication_start_date, self.page_size_field: self.page_size})
 
-        if self._replication_end_date:
+            if self._replication_start_date and self.cursor_field:
+                start_date = max(stream_state.get(self.cursor_field, self._replication_start_date), self._replication_start_date)
+                params.update({self.replication_start_date_field: start_date})
+
+        if self.replication_end_date_field is not None and self._replication_end_date:
```
Comment (marcosmarxm) [inline airbyte-integrations/connectors/source-amazon-seller-partner/source_amazon_seller_partner/streams.py:115]: Any reason why you're doing this change? Can you add a unit test to validate the new logic?

**Review by marcosmarxm — CHANGES_REQUESTED** [review]

File: airbyte-integrations/connectors/source-amazon-seller-partner/source_amazon_seller_partner/source.py (hunk 3 above)
Comment (efimmatytsin) [reply airbyte-integrations/connectors/source-amazon-seller-partner/source_amazon_seller_partner/source.py]: oh seems like, its already fixed in 1.4.1

File: airbyte-integrations/connectors/source-amazon-seller-partner/source_amazon_seller_partner/streams.py (hunk 6 above)
Comment (marcosmarxm) [reply airbyte-integrations/connectors/source-amazon-seller-partner/source_amazon_seller_partner/streams.py:115]: @efimmatytsin I'd prefer to keep the original function unchanged and you can override it for the OrderItems case.

File: airbyte-integrations/connectors/source-amazon-seller-partner/source_amazon_seller_partner/streams.py [hunk 7]
```diff
@@ -820,6 +823,60 @@ def backoff_time(self, response: requests.Response) -> Optional[float]:
...
+    
+    def parse_response(
+        self, response: requests.Response, stream_state: Mapping[str, Any] = None, stream_slice: Mapping[str, Any] = None, **kwargs
+    ) -> Iterable[Mapping]:
+        order_items_list = response.json().get(self.data_field, {})
+        if order_items_list.get(self.next_page_token_field) is None:
+            self.cached_state[self.parent_cursor_field] = stream_slice[self.parent_cursor_field]
+        for order_item in order_items_list.get(self.name, []):
+            order_item["AmazonOrderId"] = order_items_list.get('AmazonOrderId')
+            yield order_item
```
Comment (marcosmarxm) [inline airbyte-integrations/connectors/source-amazon-seller-partner/source_amazon_seller_partner/streams.py:878]: Can you add a unit test for this function?

**Review by marcosmarxm — CHANGES_REQUESTED** [review]

**Review by marcosmarxm — CHANGES_REQUESTED** [review]
The implementation looks good @efimmatytsin small nit details and almost ready to ship.

File: airbyte-integrations/connectors/source-amazon-seller-partner/source_amazon_seller_partner/streams.py [hunk 8]
```diff
@@ -820,6 +820,72 @@ def backoff_time(self, response: requests.Response) -> Optional[float]:
...
+        if next_page_token:
+            return dict(next_page_token)
+        return {}
+    
+    def stream_slices(self, stream_state: Mapping[str, Any] = None, **kwargs) -> Iterable[Optional[Mapping[str, any]]]:
+        orders = Orders(**self.stream_kwargs)
+        for order_record in orders.read_records(sync_mode=SyncMode.incremental, stream_state=stream_state):
+            self.cached_state[self.parent_cursor_field] = order_record["LastUpdateDate"]
+            self.logger.info(f"OrderItems stream slice for order {order_record['AmazonOrderId']}")
+            time.sleep(self.default_stream_slice_delay_time)
```
[the remark this answers is not in the document]
Comment (efimmatytsin) [reply airbyte-integrations/connectors/source-amazon-seller-partner/source_amazon_seller_partner/streams.py:857]: For reduce amazon rate limiter timeouts

File: airbyte-integrations/connectors/source-amazon-seller-partner/source_amazon_seller_partner/streams.py (hunk 6 above)
Comment (efimmatytsin) [reply airbyte-integrations/connectors/source-amazon-seller-partner/source_amazon_seller_partner/streams.py]: done.

File: airbyte-integrations/connectors/source-amazon-seller-partner/source_amazon_seller_partner/streams.py [hunk 9]
```diff
@@ -820,6 +820,72 @@ def backoff_time(self, response: requests.Response) -> Optional[float]:
...
+            yield {"AmazonOrderId": order_record["AmazonOrderId"], self.parent_cursor_field: order_record["LastUpdateDate"]}
```
[the remark this answers is not in the document]
Comment (efimmatytsin) [reply airbyte-integrations/connectors/source-amazon-seller-partner/source_amazon_seller_partner/streams.py]: done

File: airbyte-integrations/connectors/source-amazon-seller-partner/source_amazon_seller_partner/streams.py (hunk 7 above)
Comment (efimmatytsin) [reply airbyte-integrations/connectors/source-amazon-seller-partner/source_amazon_seller_partner/streams.py]: done

**Review by marcosmarxm — APPROVED** [review]
Amazing @efimmatytsin I'm going to publish the contribution during the day!

**Comment by efimmatytsin** [comment]
@marcosmarxm Thank you!!

**Comment by marcosmarxm** [comment]
merged at https://github.com/airbytehq/airbyte/pull/29210

**Pull request closed without merging**